### PR TITLE
ICSAAS-362 search component integration test

### PR DESCRIPTION
**Pull request by igorsimovski**
- test the component's exported json model
 - add sample content package for the integration test
 - add expected json file
 - implement a sample content install rule
 - implement sling model json exporter client

### CODE REVIEW

File: it.tests/src/main/java/com/valtech/aem/saas/it/tests/SearchComponentIT.java [hunk 1]
```diff
@@ -0,0 +1,43 @@
...
+  @Test
+  public void testSearchModelJson() throws ClientException, JSONException, IOException {
+
+    JsonNode jsonNode = slingModelJsonExporterClient.doGetModelJson(
+        "/content/saas-aem-module/us/en/jcr:content/root/container/search", 200);
+    String expectedJson = IOUtils.toString(ResourceUtil.getResourceAsStream(EXPECTED_SEARCH_MODEL_JSON),
+        StandardCharsets.UTF_8.name());
+    JSONAssert.assertEquals(expectedJson, jsonNode.toString(), false);
+  }
+
```
Comment (Valtech-Switzerland) [inline it.tests/src/main/java/com/valtech/aem/saas/it/tests/SearchComponentIT.java:42]: Please add some more testcases (smoke tests)

- [ ] Perform a Search query, check if results are returned
- [ ] Check if results differ between the tabs
- [ ] Check if facet items will be returned
- [ ] Check autocomplete
- [ ] Check autsuggest

File: it.tests/src/main/java/com/valtech/aem/saas/it/tests/SearchComponentIT.java [hunk 2]
```diff
@@ -0,0 +1,145 @@
...
+    JsonNode jsonNode = slingModelJsonExporterClient.doGetJsonNode(
+        getJsonModelPreparedUrl("/content/saas-aem-module/us/en/jcr:content/root/container/search"), 200);
+    String expectedJson = IOUtils.toString(ResourceUtil.getResourceAsStream(EXPECTED_SEARCH_MODEL_JSON),
+        StandardCharsets.UTF_8.name());
+    JSONAssert.assertEquals(expectedJson, jsonNode.toString(), false);
+  }
+
+  @Test
+  public void testSearchResults() throws ClientException {
+    List<NameValuePair> searchParams = Arrays.asList(new BasicNameValuePair("q", "abb"),
```
Comment (nhirrle) [inline it.tests/src/main/java/com/valtech/aem/saas/it/tests/SearchComponentIT.java:55]: please change query-term to something neutral ;)

File: it.tests/src/main/java/com/valtech/aem/saas/it/tests/SearchComponentIT.java [hunk 3]
```diff
@@ -0,0 +1,145 @@
...
+        new BasicNameValuePair("lang", "en"));
```
Comment (nhirrle) [inline it.tests/src/main/java/com/valtech/aem/saas/it/tests/SearchComponentIT.java:56]: for q and lang, please use the contants defined in the module.

**Review by nhirrle — COMMENTED** [review]
cool intergration test, hope this will help in the future :)

File: it.tests/src/main/java/com/valtech/aem/saas/it/tests/SearchComponentIT.java [hunk 4]
```diff
@@ -0,0 +1,147 @@
...
+            "/content/saas-aem-module/us/en/jcr:content/root/container/search/search-tabs/searchtab"), searchParams,
+        200);
+
+    validateSearchTabJson(searchTab1);
+    assertEquals("food", getSuggestionText(searchTab1));
+
+  }
+
+  private String getSuggestionText(JsonNode searchTab) {
+    JsonNode suggestion = searchTab.get("suggestion");
```
Comment (nhirrle) [inline it.tests/src/main/java/com/valtech/aem/saas/it/tests/SearchComponentIT.java:108]: - [ ] also this var please - use a constant if possible

File: it.tests/src/main/java/com/valtech/aem/saas/it/tests/SearchComponentIT.java [hunk 5]
```diff
@@ -0,0 +1,147 @@
...
+    assertTrue(suggestion.isObject());
+    JsonNode text = suggestion.get("text");
```
Comment (nhirrle) [inline it.tests/src/main/java/com/valtech/aem/saas/it/tests/SearchComponentIT.java:110]: - [ ]  also this var please - use a constant if possible

File: it.tests/src/main/java/com/valtech/aem/saas/it/tests/SearchComponentIT.java [hunk 6]
```diff
@@ -0,0 +1,147 @@
...
+    assertNotNull(text);
+    assertTrue(text.isTextual());
+    return text.getTextValue();
+  }
+
+  private void hasFacetFilters(JsonNode searchTab) {
+    JsonNode facetFilters = searchTab.get("facetFilters");
```
Comment (nhirrle) [inline it.tests/src/main/java/com/valtech/aem/saas/it/tests/SearchComponentIT.java:117]: - [ ]  also this var please - use a constant if possible

File: it.tests/src/main/java/com/valtech/aem/saas/it/tests/SearchComponentIT.java [hunk 7]
```diff
@@ -0,0 +1,147 @@
...
+    assertNotNull(facetFilters);
+    assertTrue(facetFilters.isObject());
+    JsonNode items = facetFilters.get("items");
```
Comment (nhirrle) [inline it.tests/src/main/java/com/valtech/aem/saas/it/tests/SearchComponentIT.java:120]: - [ ]  also this var please - use a constant if possible

File: it.tests/src/main/java/com/valtech/aem/saas/it/tests/SearchComponentIT.java [hunk 8]
```diff
@@ -0,0 +1,147 @@
...
+    return resultsTotal.getIntValue();
+  }
+
+  private void validateSearchTabJson(JsonNode searchTab) {
+    assertNotNull(searchTab);
+    assertTrue(searchTab.isObject());
+  }
+
+  private void searchResultsExist(JsonNode searchTab) {
+    JsonNode results = searchTab.get("results");
```
Comment (nhirrle) [inline it.tests/src/main/java/com/valtech/aem/saas/it/tests/SearchComponentIT.java:138]: - [ ]  also this var please - use a constant if possible

File: it.tests/pom.xml [hunk 9]
```diff
@@ -35,6 +35,16 @@
...
 
     <dependencies>
+        <dependency>
+            <groupId>io.github.valtech-ch</groupId>
+            <artifactId>saas-aem.api</artifactId>
+            <version>${project.version}</version>
+        </dependency>
+        <dependency>
+            <groupId>io.github.valtech-ch</groupId>
+            <artifactId>saas-aem.core</artifactId>
```
Comment (nhirrle) [inline it.tests/pom.xml:45]: why do we need core dependency?

**Review by nhirrle — APPROVED** [review]

File: it.tests/pom.xml (hunk 9 above)
Comment (igorsimovski) [reply it.tests/pom.xml:45]: to be able to use the constant SearchTabModelImpl.SEARCH_TERM

Comment (nhirrle) [reply it.tests/pom.xml:45]: OK, please move this to the api. It's clearly something the implementer needs to know

**Pull request merged**